Cannot inject @Test annotated Method [main] with [class [Ljava.lang.String;].

```java
@Test(singleThreaded = true)
public class MainTest {

  @Test public void foo() {}

  public static void main(String[] args) {}
}
```

```
[Utils] [ERROR] [Error] org.testng.TestNGException: 
Cannot inject @Test annotated Method [main] with [class [Ljava.lang.String;].
```

While understandable, I'd argue that the class annotation shouldn't apply to static methods or, at least, to the `main` method. It is easy to work around by configuring each method, but was surprising.

I think it's because of the fact that a class level @Test annotation causes
TestNG to start treating all public methods (both static and instance ones)
as Test methods and so tries parameter injection (for parameterized methods
) which explains the error because default parameter injection doesn't know
how to deal with a String array.

On Sun, Apr 2, 2017, 08:09 Ben Manes <notifications@github.com> wrote:

> @Test(singleThreaded = true)public class MainTest {
>
>   @Test public void foo() {}
>
>   public static void main(String[] args) {}
> }
>
> [Utils] [ERROR] [Error] org.testng.TestNGException:
> Cannot inject @Test annotated Method [main] with [class [Ljava.lang.String;].
>
> While understandable, I'd argue that the class annotation shouldn't apply
> to static methods or, at least, to the main method. It is easy to work
> around by configuring each method, but was surprising.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cbeust/testng/issues/1405>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3hzexRw63Rm-DtQl-EIl3-6hdbuOm_ks5rrwpQgaJpZM4Mwrba>
> .
>
-- 

Thanks & Regards
Krishnan Mahadevan

"All the desirable things in life are either illegal, expensive, fattening
or in love with someone else!"
My Scribblings @ http://wakened-cognition.blogspot.com/

My Technical Scribbings @ http://rationaleemotions.wordpress.com/

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-1405: Skip considering main() method when @Test used at class level (Krishnan Mahadevan)
 Fixed: GITHUB-799: @Factory with dataProvider changes order of iterations (Krishnan Mahadevan & Julien Herr)
 New: Enhance XML Reporter to be able to customize the file name (Krishnan Mahadevan)
 Fixed: GITHUB-1417: Class param injection is not working with @BeforeClass (Krishnan Mahadevan)

--- a/src/main/java/org/testng/internal/reflect/ReflectionHelper.java
+++ b/src/main/java/org/testng/internal/reflect/ReflectionHelper.java
@@ -15,7 +15,7 @@ public class ReflectionHelper {
    */
   public static Method[] getLocalMethods(Class<?> clazz) {
     Method[] result;
-    Method[] declaredMethods = clazz.getDeclaredMethods();
+    Method[] declaredMethods = excludingMain(clazz);
     List<Method> defaultMethods = getDefaultMethods(clazz);
     if (defaultMethods != null) {
       result = new Method[declaredMethods.length + defaultMethods.size()];
@@ -36,6 +36,36 @@ public class ReflectionHelper {
       result = prunedMethods.toArray(new Method[prunedMethods.size()]);
     }
     return result;
+  }
+
+  /**
+   * @return An array of all locally declared methods or equivalent thereof
+   * (such as default methods on Java 8 based interfaces that the given class
+   * implements) but excludes the <code>main()</code> method alone.
+   */
+  public static Method[] excludingMain(Class<?> clazz) {
+    Method[] declaredMethods = clazz.getDeclaredMethods();
+    List<Method> pruned = new LinkedList<>();
+    for (Method declaredMethod :declaredMethods) {
+      if ("main".equals(declaredMethod.getName()) && isStaticVoid(declaredMethod) && acceptsStringArray(declaredMethod)) {
+        continue;
+      }
+      pruned.add(declaredMethod);
+    }
+    return pruned.toArray(new Method[0]);
+  }
+
+  private static boolean isStaticVoid(Method method) {
+    return method.getReturnType().equals(void.class) && Modifier.isStatic(method.getModifiers());
+  }
+
+  private static boolean acceptsStringArray(Method method) {
+    Class<?>[] paramTypes = method.getParameterTypes();
+    if (paramTypes.length == 0) {
+      return false;
+    }
+    Class<?> paramType = paramTypes[0];
+    return paramType.isArray() && paramType.isInstance(new String[0]);
   }
 
   private static List<Method> getDefaultMethods(Class<?> clazz) {

--- a/src/main/java/org/testng/internal/reflect/ReflectionHelper.java
+++ b/src/main/java/org/testng/internal/reflect/ReflectionHelper.java
@@ -52,7 +52,7 @@ public class ReflectionHelper {
       }
       pruned.add(declaredMethod);
     }
-    return pruned.toArray(new Method[0]);
+    return pruned.toArray(new Method[pruned.size()]);
   }
 
   private static boolean isStaticVoid(Method method) {

--- a/src/main/java/org/testng/junit/JUnitMethodFinder.java
+++ b/src/main/java/org/testng/junit/JUnitMethodFinder.java
@@ -6,6 +6,7 @@ import org.testng.collections.Lists;
 import org.testng.internal.ConstructorOrMethod;
 import org.testng.internal.TestNGMethod;
 import org.testng.internal.annotations.IAnnotationFinder;
+import org.testng.internal.reflect.ReflectionHelper;
 import org.testng.xml.XmlTest;
 
 import java.lang.reflect.Constructor;
@@ -80,7 +81,7 @@ public class JUnitMethodFinder implements ITestMethodFinder {
     //
     Class current = cls;
     while(!(current == Object.class)) {
-      Method[] allMethods = current.getDeclaredMethods();
+      Method[] allMethods = ReflectionHelper.excludingMain(current);
       for(Method allMethod : allMethods) {
         ITestNGMethod m = new TestNGMethod(/* allMethods[i].getDeclaringClass(), */ allMethod,
             m_annotationFinder, null,

--- a/src/test/java/test/ReflectionHelperTest.java
+++ b/src/test/java/test/ReflectionHelperTest.java
@@ -4,6 +4,7 @@ import org.testng.Assert;
 import org.testng.annotations.Test;
 import org.testng.collections.Lists;
 import org.testng.internal.reflect.ReflectionHelper;
+import test.github1405.TestClassSample;
 import test.github765.DuplicateCallsSample;
 
 import java.lang.reflect.Method;
@@ -28,6 +29,10 @@ public class ReflectionHelperTest {
         //Abstract methods should be included.
         methods = prune(ReflectionHelper.getLocalMethods(Dragon.class));
         Assert.assertEquals(methods.length, 2);
+
+        //main methods should be pruned
+        methods = prune(ReflectionHelper.getLocalMethods(TestClassSample.class));
+        Assert.assertEquals(methods.length, 1);
     }
 
     /**

--- a/src/test/java/test/github1405/JUnitTestClassSample.java
+++ b/src/test/java/test/github1405/JUnitTestClassSample.java
@@ -1,0 +1,12 @@
+package test.github1405;
+
+import org.junit.Test;
+
+public class JUnitTestClassSample {
+    @Test
+    public void testMethod() {}
+    @Test
+    public static void main(String[] args) {
+
+    }
+}

--- a/src/test/java/test/github1405/TestClassSample.java
+++ b/src/test/java/test/github1405/TestClassSample.java
@@ -1,0 +1,14 @@
+package test.github1405;
+
+
+import org.testng.annotations.Test;
+
+@Test
+public class TestClassSample {
+    public static void main(String[] args) {
+    }
+
+    public void testMethod() {
+    }
+
+}

--- a/src/test/java/test/github1405/TestExclusionOfMainMethod.java
+++ b/src/test/java/test/github1405/TestExclusionOfMainMethod.java
@@ -21,8 +21,7 @@ public class TestExclusionOfMainMethod extends SimpleBaseTest {
         XmlSuite xmlSuite = createXmlSuite("suite");
         xmlSuite.setJunit(true);
         createXmlTest(xmlSuite, "test", JUnitTestClassSample.class);
-        TestNG tng = create();
-        tng.setXmlSuites(Collections.singletonList(xmlSuite));
+        TestNG tng = create(xmlSuite);
         tng.run();
         Assert.assertEquals(tng.getStatus(), 0);
     }

--- a/src/test/java/test/github1405/TestExclusionOfMainMethod.java
+++ b/src/test/java/test/github1405/TestExclusionOfMainMethod.java
@@ -1,0 +1,29 @@
+package test.github1405;
+
+import org.testng.Assert;
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlSuite;
+import test.SimpleBaseTest;
+
+import java.util.Collections;
+
+public class TestExclusionOfMainMethod extends SimpleBaseTest {
+    @Test
+    public void testMainMethodExclusion() {
+        TestNG tng = create(TestClassSample.class);
+        tng.run();
+        Assert.assertEquals(tng.getStatus(), 0);
+    }
+
+    @Test
+    public void testMainMethodExclusionForJunit() {
+        XmlSuite xmlSuite = createXmlSuite("suite");
+        xmlSuite.setJunit(true);
+        createXmlTest(xmlSuite, "test", JUnitTestClassSample.class);
+        TestNG tng = create();
+        tng.setXmlSuites(Collections.singletonList(xmlSuite));
+        tng.run();
+        Assert.assertEquals(tng.getStatus(), 0);
+    }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -97,6 +97,7 @@
       <class name="test.thread.MultiThreadedDependentTest" />
       <class name="test.thread.TrueParallelTest" />
       <class name="test.github765.ExcludeSyntheticMethodsFromTemplateCallsTest"/>
+      <class name="test.github1405.TestExclusionOfMainMethod"/>
 <!--
       <class name="test.thread.ParallelSuiteTest"/>
 -->

